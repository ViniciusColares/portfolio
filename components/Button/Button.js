import React from "react";
import styled from "styled-components";
import css from "@styled-system/css";
import PropTypes from "prop-types";
import { variant, space } from "styled-system";

const Button = ({ name, variant, action, type, icon, ...restProps }) => {
  return (
    <ButtonStyle variant={variant} onClick={action} type={type} {...restProps}>
      {icon && <img alt={name} src={icon} />}
      {name}
    </ButtonStyle>
  );
};

const ButtonStyle = styled("button")(
  css({
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    minWidth: "135px",
    whiteSpace: "nowrap",
    fontFamily: "text",
    bg: "accent",
    padding: "8px 25px",
    fontSize: 1,
    fontWeight: 600,
    color: "primaryDark",
    textAlign: "center",
    borderRadius: "40px",
    cursor: "pointer",
    border: "none",
    outline: "none",
    boxShadow: "0px 4px 4px rgba(0, 0, 0, 0.35)",
    "&: hover": {
      bg: "accentDark",
    },
    "> img": {
      marginRight: 2,
    },
  }),
  space,
  variant({
    variants: {
      small: {
        minWidth: "auto",
        padding: "5px 16px",
        fontSize: 0,
      },
      big: {
        minWidth: "180px",
        padding: "20px 40px",
        fontSize: 3,
      },
      disabled: {
        border: "none",
        background: "none",
        backgroundColor: "grayDefaultColor",
        color: "blackLight",
        cursor: "not-allowed",
        "&:hover": {
          border: "none",
          color: "inherit",
          background: "none",
          bg: "grayDefaultColor",
        },
      },
      transparent: {
        background: "none",
        backgroundColor: "transparent",
        color: "defaultColor",
        "&:hover": {
          backgroundColor: "focusInput",
        },
      },
    },
  })
);

Button.defaultProps = {
  type: "button",
  name: "Custom Button",
  variant: [],
  isLoading: false,
};

Button.propTypes = {
  type: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  variant: PropTypes.arrayOf(PropTypes.string),
  action: PropTypes.func,
  isLoading: PropTypes.bool,
};

export default Button;
