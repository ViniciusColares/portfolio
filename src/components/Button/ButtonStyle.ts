import styled from 'styled-components'
import css from '@styled-system/css'
import { variant } from 'styled-system'
import { theme } from '@styles/theme'
import { motion } from 'framer-motion'

const { colors } = theme

export const Container = styled.div(
  css({
    position: 'relative'
  })
)

export const ButtonStyle = styled(motion.button)(
  css({
    fontSize: 1,
    bg: 'accent',
    display: 'flex',
    outline: 'none',
    fontWeight: 600,
    cursor: 'pointer',
    userSelect: 'none',
    fontFamily: 'text',
    borderWidth: '1px',
    padding: '8px 20px',
    textAlign: 'center',
    borderRadius: '8px',
    color: 'primaryDark',
    position: 'relative',
    alignItems: 'center',
    whiteSpace: 'nowrap',
    borderStyle: 'solid',
    justifyContent: 'center',
    transition: 'all .15s ease',
    borderColor: 'accentDark',
    '&:focus': {
      boxShadow: `inset 0px 0px 0px 3px ${colors.accentDark}`,
      '&:hover': {
        transform: 'translateX(2px) translateY(2px)'
      }
    },
    '&:hover': {
      transform: 'translateX(2px) translateY(2px)',
      '&:active': {
        transform: 'translateX(4px) translateY(4px)'
      }
    },
    '&:active': {
      transform: 'translateX(4px) translateY(4px)'
    },
    '&:disabled': {
      borderColor: 'gray50',
      bg: 'gray25',
      color: 'gray50',
      cursor: 'not-allowed',
      '&:hover': {
        transform: 'translateY(0px)'
      },
      '&+div': {
        visibility: 'hidden'
      }
    },
    '> img': {
      marginRight: 2
    }
  }),
  variant({
    prop: 'size',
    variants: {
      small: {
        minWidth: 'auto',
        padding: '4px 12px',
        fontSize: 0,
        fontWeight: 600,
        '&:focus': {
          boxShadow: `inset 0px 0px 0px 2px ${colors.accentDark}`,
          '&:hover': {
            transform: 'translateX(2px) translateY(2px)'
          }
        },
        '&:hover': {
          transform: 'translateX(2px) translateY(2px)',
          '&:active': {
            transform: 'translateX(4px) translateY(4px)'
          }
        },
        '&:active': {
          transform: 'translateX(4px) translateY(4px)'
        },
        '> img': {
          marginRight: 1
        }
      },
      big: {
        minWidth: '150px',
        padding: '12px 32px',
        borderWidth: '2px',
        fontSize: 2,
        fontWeight: 600,
        '&:focus': {
          boxShadow: `inset 0px 0px 0px 4px ${colors.accentDark}`,
          '&:hover': {
            transform: 'translateX(2px) translateY(2px)'
          }
        },
        '&:hover': {
          transform: 'translateX(2px) translateY(2px)',
          '&:active': {
            transform: 'translateX(5px) translateY(5px)'
          }
        },
        '&:active': {
          transform: 'translateX(5px) translateY(5px)'
        },
        '> img': {
          marginRight: 3
        }
      }
    }
  })
)

export const Shadow = styled.div(
  css({
    position: 'absolute',
    width: '100%',
    height: '100%',
    bg: 'shadow',
    borderRadius: '8px',
    border: `1px solid ${colors.shadow}`,
    transform: 'translateY(4px) translateX(4px)',
    top: 0,
    left: 0,
    zIndex: -1
  }),
  variant({
    prop: 'size',
    variants: {
      small: {
        transform: 'translateY(3px) translateX(3px)'
      },
      big: {
        transform: 'translateY(5px) translateX(5px)'
      }
    }
  })
)
