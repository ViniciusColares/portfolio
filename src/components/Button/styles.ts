import styled from 'styled-components'
import css from '@styled-system/css'
import { variant, compose, space } from 'styled-system'
import { colors } from '@styles/theme'
import { motion, MotionProps } from 'framer-motion'

type ButtonStyleProps = { size?: string }

export const Container = styled.div(
  css({
    position: 'relative'
  })
)

export const ButtonStyle = styled(motion.button)<
  MotionProps & ButtonStyleProps
>(
  css({
    fontSize: 2,
    bg: 'accent',
    display: 'flex',
    outline: 'none',
    fontWeight: 600,
    cursor: 'pointer',
    userSelect: 'none',
    fontFamily: 'text',
    borderWidth: '1px',
    padding: '8px 20px',
    textAlign: 'center',
    borderRadius: '8px',
    color: 'primaryDark',
    position: 'relative',
    alignItems: 'center',
    whiteSpace: 'nowrap',
    borderStyle: 'solid',
    justifyContent: 'center',
    transition: 'all .15s ease',
    borderColor: 'accentDark',
    transform: 'translateY(-4px) translateX(-4px)',
    '&:focus': {
      boxShadow: `inset 0px 0px 0px 3px ${colors.accentDark}`,
      '&:hover': {
        transform: 'translateX-(2px) translateY(-2px)'
      }
    },
    '&:hover': {
      transform: 'translateX(-2px) translateY(-2px)',
      '&:active': {
        transform: 'translateX(0px) translateY(0px)'
      }
    },
    '&:active': {
      transform: 'translateX(0px) translateY(0px)'
    },
    '&:disabled': {
      borderColor: 'gray50',
      bg: 'gray25',
      color: 'gray50',
      cursor: 'not-allowed',
      '&:hover': {
        transform: 'translateY(0px)'
      },
      '&+div': {
        visibility: 'hidden'
      }
    },
    '> svg': {
      marginRight: 2
    }
  }),
  variant({
    prop: 'size',
    variants: {
      small: {
        minWidth: 'auto',
        padding: '4px 12px',
        fontSize: 1,
        fontWeight: 600,
        transform: 'translateY(-3px) translateX(-3px)',
        '&:focus': {
          boxShadow: `inset 0px 0px 0px 2px ${colors.accentDark}`,
          '&:hover': {
            transform: 'translateX(-2px) translateY(-2px)'
          }
        },
        '&:hover': {
          transform: 'translateX(-2px) translateY(-2px)',
          '&:active': {
            transform: 'translateX(0px) translateY(0px)'
          }
        },
        '&:active': {
          transform: 'translateX(0px) translateY(0px)'
        },
        '> svg': {
          marginRight: 1
        }
      },
      big: {
        minWidth: '150px',
        padding: '12px 32px',
        borderWidth: '2px',
        fontSize: 3,
        fontWeight: 600,
        transform: 'translateY(-5px) translateX(-5px)',
        '&:focus': {
          boxShadow: `inset 0px 0px 0px 4px ${colors.accentDark}`,
          '&:hover': {
            transform: 'translateX(-3px) translateY(-3px)'
          }
        },
        '&:hover': {
          transform: 'translateX(-3px) translateY(-3px)',
          '&:active': {
            transform: 'translateX(0px) translateY(0px)'
          }
        },
        '&:active': {
          transform: 'translateX(0px) translateY(0px)'
        },
        '> svg': {
          marginRight: 3
        }
      }
    }
  }),
  compose(space)
)

export const Shadow = styled.div(
  css({
    position: 'absolute',
    width: '100%',
    height: '100%',
    bg: 'shadow',
    borderRadius: '8px',
    border: `1px solid ${colors.shadow}`,
    top: 0,
    left: 0,
    zIndex: -1
  })
)
