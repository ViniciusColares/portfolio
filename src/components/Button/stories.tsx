import React, { ComponentProps } from 'react'
import { Story, Meta } from '@storybook/react/types-6-0'

import Button from '.'

export default {
  title: 'Core/Button',
  component: Button
} as Meta

const Template: Story<ComponentProps<typeof Button>> = (args) => (
  <Button {...args} />
)

export const Basic = Template.bind({})
Basic.args = {
  name: 'Button'
}
Basic.argTypes = { onClick: { action: 'clicked' } }

export const Loading = Template.bind({})
Loading.args = {
  name: 'Loading',
  isLoading: true
}
