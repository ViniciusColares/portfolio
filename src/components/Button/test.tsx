import { render, screen } from '@testing-library/react'

import Button from './Button'

describe('<Button />', () => {
  it('it should render text inside button', () => {
    const { container } = render(<Button name="botão" />)
    expect(screen.getByText('botão')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('it should render a loading button', () => {
    const { container } = render(<Button name="botão" isLoading />)
    expect(screen.getByText('carregando')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
