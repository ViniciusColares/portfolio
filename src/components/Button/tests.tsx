import { render, screen } from '@testing-library/react'

import Button from '.'

describe('Button', () => {
  it('it should render text inside button', () => {
    const { container } = render(<Button name="botão" />)
    expect(screen.getByText('botão')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('it should render a loading button', () => {
    const { container } = render(
      <Button name="botão" isLoading data-testid="svgButton" />
    )
    expect(screen.getByTestId('svgButton')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
})
