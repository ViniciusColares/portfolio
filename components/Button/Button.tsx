import React from "react";

import { ButtonStyle } from "./ButtonStyle";

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: string | string[];
  isLoading: boolean;
  icon?: string;
  size?: string;
  state?: string;
  fill?: string;
}

const Button = ({ name, type, icon, isLoading, ...restProps }: ButtonProps) => (
  <ButtonStyle type={type} {...restProps}>
    {isLoading ? (
      "carregando"
    ) : (
      <>
        {icon && <img alt={name} src={icon} />}
        {name}
      </>
    )}
  </ButtonStyle>
);

Button.defaultProps = {
  type: "button",
  isLoading: false,
};

export default Button;
