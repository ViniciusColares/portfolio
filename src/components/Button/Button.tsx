import React from 'react'
import { IconBaseProps } from 'react-icons'
import { VscLoading } from 'react-icons/vsc'
import { Rotate } from '@styles/animations'

import { ButtonStyle } from './ButtonStyle'

const animations = {
  initial: {
    width: 'auto'
  }
}

const Button = ({
  name,
  type = 'button',
  icon: Icon,
  isLoading = false,
  ...restProps
}: ButtonProps) => {
  return (
    <ButtonStyle
      animate={'initial'}
      variants={animations}
      transition={{ type: 'spring' }}
      type={type}
      disabled={isLoading}
      {...restProps}
    >
      {isLoading ? (
        <Rotate
          animate={{ rotate: 360 }}
          transition={{
            loop: Infinity,
            ease: 'linear',
            duration: 1
          }}
        >
          <VscLoading size="23px" color="#938CA3" />
        </Rotate>
      ) : (
        <>
          {Icon && <Icon size="24px" color="#fff" />}
          {name}
        </>
      )}
    </ButtonStyle>
  )
}

export interface ButtonProps {
  type: 'button' | 'submit' | 'reset'
  name: string
  isLoading?: boolean
  size?: 'small' | 'big'
  icon?: React.ComponentType<IconBaseProps>
}

export default Button
