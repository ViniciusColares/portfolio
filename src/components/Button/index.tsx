import React from 'react'
import { IconBaseProps } from 'react-icons'
import { VscLoading } from 'react-icons/vsc'
import { MotionProps } from 'framer-motion'
import { SpaceProps } from 'styled-system'

import { Rotate } from '@styles/animations'
import { colors } from '@styles/theme'

import * as $ from './styles'

const Button = ({
  name,
  type = 'button',
  icon: Icon,
  size = 'medium',
  isLoading = false,
  ...restProps
}: ButtonProps &
  React.HTMLAttributes<HTMLButtonElement> &
  MotionProps &
  SpaceProps) => {
  return (
    <$.Container>
      <$.ButtonStyle
        type={type}
        size={size}
        disabled={isLoading}
        {...restProps}
      >
        {isLoading ? (
          <Rotate
            animate={{ rotate: 360 }}
            transition={{
              loop: Infinity,
              ease: 'linear',
              duration: 1
            }}
          >
            <VscLoading size="23px" color="#938CA3" />
          </Rotate>
        ) : (
          <>
            {Icon && <Icon size="24px" color={colors.primaryDark} />}
            {name}
          </>
        )}
      </$.ButtonStyle>
      <$.Shadow />
    </$.Container>
  )
}

export interface ButtonProps {
  name: string
  type?: 'button' | 'submit' | 'reset'
  onClick?: (() => Promise<boolean>) | (() => void)
  isLoading?: boolean
  size?: 'small' | 'medium' | 'big'
  icon?: React.ComponentType<IconBaseProps>
}

export default Button
