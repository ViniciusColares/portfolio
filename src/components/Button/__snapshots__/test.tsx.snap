// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button it should render a loading button 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
  font-family: text;
  background-color: accent;
  font-size: 14px;
  font-weight: 600;
  color: primaryDark;
  text-align: center;
  border-radius: 6px;
  cursor: pointer;
  border: none;
  outline: none;
  -webkit-transition: all .3s ease-out;
  transition: all .3s ease-out;
}

.c0:focus {
  box-shadow: inset 0 0 0px 4px #BB8B0E;
}

.c0:hover {
  background-color: #d69b00;
  box-shadow: 0px 0px 6px rgba(170,140,210,0.9);
}

.c0:disabled {
  background-color: gray25;
  color: gray50;
  cursor: not-allowed;
}

.c0 > img {
  margin-right: 8px;
}

<button
  class="c0"
  data-testid="svgButton"
  disabled=""
  style="width: 0px;"
  type="button"
>
  <div
    class="c1"
  >
    <svg
      color="#938CA3"
      fill="currentColor"
      height="23px"
      stroke="currentColor"
      stroke-width="0"
      style="color: rgb(147, 140, 163);"
      viewBox="0 0 16 16"
      width="23px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 .75c.172 0 .333.034.484.102a1.214 1.214 0 0 1 .664.664c.068.15.102.312.102.484s-.034.333-.102.484a1.214 1.214 0 0 1-.265.399 1.324 1.324 0 0 1-.399.273A1.254 1.254 0 0 1 8 3.25c-.172 0-.333-.031-.484-.094a1.324 1.324 0 0 1-.672-.672A1.254 1.254 0 0 1 6.75 2c0-.172.031-.333.094-.484.067-.151.159-.284.273-.399.115-.114.248-.203.399-.265A1.17 1.17 0 0 1 8 .75zM2.633 3.758a1.111 1.111 0 0 1 .68-1.031 1.084 1.084 0 0 1 .882 0c.136.057.253.138.352.242.104.099.185.216.242.351a1.084 1.084 0 0 1 0 .883 1.122 1.122 0 0 1-.594.594 1.169 1.169 0 0 1-.883 0 1.19 1.19 0 0 1-.359-.234 1.19 1.19 0 0 1-.234-.36 1.169 1.169 0 0 1-.086-.445zM2 7a.941.941 0 0 1 .703.297A.941.941 0 0 1 3 8a.97.97 0 0 1-.078.39 1.03 1.03 0 0 1-.531.532A.97.97 0 0 1 2 9a.97.97 0 0 1-.39-.078 1.104 1.104 0 0 1-.32-.211 1.104 1.104 0 0 1-.212-.32A.97.97 0 0 1 1 8a.97.97 0 0 1 .29-.703A.97.97 0 0 1 2 7zm.883 5.242a.887.887 0 0 1 .531-.805.863.863 0 0 1 .68 0c.11.047.203.11.281.188a.887.887 0 0 1 .188.96.887.887 0 0 1-1.148.461.913.913 0 0 1-.462-.46.863.863 0 0 1-.07-.344zM8 13.25c.208 0 .385.073.531.219A.723.723 0 0 1 8.75 14a.723.723 0 0 1-.219.531.723.723 0 0 1-.531.219.723.723 0 0 1-.531-.219A.723.723 0 0 1 7.25 14c0-.208.073-.385.219-.531A.723.723 0 0 1 8 13.25zm3.617-1.008c0-.177.06-.325.18-.445s.268-.18.445-.18.326.06.445.18c.12.12.18.268.18.445s-.06.326-.18.445a.605.605 0 0 1-.445.18.605.605 0 0 1-.445-.18.605.605 0 0 1-.18-.445zM14 7.5a.48.48 0 0 1 .352.148A.48.48 0 0 1 14.5 8a.48.48 0 0 1-.148.352A.48.48 0 0 1 14 8.5a.48.48 0 0 1-.352-.148A.48.48 0 0 1 13.5 8a.48.48 0 0 1 .148-.352A.48.48 0 0 1 14 7.5zm-1.758-5.117c.188 0 .365.036.531.11a1.413 1.413 0 0 1 .735.734c.073.166.11.343.11.53 0 .188-.037.365-.11.532a1.413 1.413 0 0 1-.735.734 1.31 1.31 0 0 1-.53.11c-.188 0-.365-.037-.532-.11a1.415 1.415 0 0 1-.734-.734 1.31 1.31 0 0 1-.11-.531c0-.188.037-.365.11-.531a1.413 1.413 0 0 1 .734-.735c.167-.073.344-.11.531-.11z"
      />
    </svg>
  </div>
</button>
`;

exports[`Button it should render text inside button 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
  font-family: text;
  background-color: accent;
  font-size: 14px;
  font-weight: 600;
  color: primaryDark;
  text-align: center;
  border-radius: 6px;
  cursor: pointer;
  border: none;
  outline: none;
  -webkit-transition: all .3s ease-out;
  transition: all .3s ease-out;
}

.c0:focus {
  box-shadow: inset 0 0 0px 4px #BB8B0E;
}

.c0:hover {
  background-color: #d69b00;
  box-shadow: 0px 0px 6px rgba(170,140,210,0.9);
}

.c0:disabled {
  background-color: gray25;
  color: gray50;
  cursor: not-allowed;
}

.c0 > img {
  margin-right: 8px;
}

<button
  class="c0"
  style="width: 0px;"
  type="button"
>
  bot√£o
</button>
`;
